File: app/src/main/java/com/ph03nix_x/capacityinfo/view/CenteredToolbar.java
package com.ph03nix_x.capacityinfo.view;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.Gravity;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.Toolbar;

import com.ph03nix_x.capacityinfo.R;

public class CenteredToolbar extends Toolbar {

    private TextView tvTitle;
    private TextView tvSubtitle;

    private LinearLayout linear;

    public CenteredToolbar(Context context) {
        super(context);
        setupTextViews();
    }

    public CenteredToolbar(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setupTextViews();
    }

    public CenteredToolbar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setupTextViews();
    }

    @Override
    public CharSequence getTitle() {
        return tvTitle.getText().toString();
    }

    @Override
    public void setTitle(@StringRes int resId) {
        String s = getResources().getString(resId);
        setTitle(s);
    }

    @Override
    public void setTitle(CharSequence title) {
        tvTitle.setText(title);
    }

    @Override
    public CharSequence getSubtitle() {
        return tvSubtitle.getText().toString();
    }

    @Override
    public void setSubtitle(int resId) {
        String s = getResources().getString(resId);
        setSubtitle(s);
    }

    @Override
    public void setSubtitle(CharSequence subtitle) {
        tvSubtitle.setVisibility(VISIBLE);
        tvSubtitle.setText(subtitle);
    }

    private void setupTextViews() {
        tvSubtitle = new TextView(getContext());
        tvTitle = new TextView(getContext());

        tvTitle.setEllipsize(TextUtils.TruncateAt.END);
        tvTitle.setTextAppearance(getContext(), R.style.TextAppearance_AppCompat_Widget_ActionBar_Title);

        linear = new LinearLayout(getContext());
        linear.setGravity(Gravity.CENTER);
        linear.setOrientation(LinearLayout.VERTICAL);
        linear.addView(tvTitle);
        linear.addView(tvSubtitle);

        tvSubtitle.setSingleLine();
        tvSubtitle.setEllipsize(TextUtils.TruncateAt.END);
        tvSubtitle.setTextAppearance(getContext(), R.style.TextAppearance_AppCompat_Widget_ActionBar_Subtitle);

        tvSubtitle.setVisibility(GONE);

        LayoutParams lp = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        lp.gravity = Gravity.CENTER;
        linear.setLayoutParams(lp);

        addView(linear);
    }

}

File: app/src/main/java/com/ph03nix_x/capacityinfo/view/CenteredToolbar.kt
package com.ph03nix_x.capacityinfo.view

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.appcompat.widget.Toolbar

import com.ph03nix_x.capacityinfo.R

class CenteredToolbar : Toolbar {

    private var tvTitle: TextView? = null
    private var tvSubtitle: TextView? = null

    private var linear: LinearLayout? = null

    constructor(context: Context) : super(context) {
        setupTextViews()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        setupTextViews()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        setupTextViews()
    }

    override fun getTitle(): CharSequence {
        return tvTitle!!.text.toString()
    }

    override fun setTitle(@StringRes resId: Int) {
        val s = resources.getString(resId)
        title = s
    }

    override fun setTitle(title: CharSequence) {
        tvTitle!!.text = title
    }

    override fun getSubtitle(): CharSequence {
        return tvSubtitle!!.text.toString()
    }

    override fun setSubtitle(resId: Int) {
        val s = resources.getString(resId)
        subtitle = s
    }

    override fun setSubtitle(subtitle: CharSequence) {
        tvSubtitle!!.visibility = View.VISIBLE
        tvSubtitle!!.text = subtitle
    }

    private fun setupTextViews() {
        tvSubtitle = TextView(context)
        tvTitle = TextView(context)

        tvTitle!!.ellipsize = TextUtils.TruncateAt.END
        tvTitle!!.setTextAppearance(
            context,
            R.style.TextAppearance_AppCompat_Widget_ActionBar_Title
        )

        linear = LinearLayout(context)
        linear!!.gravity = Gravity.CENTER
        linear!!.orientation = LinearLayout.VERTICAL
        linear!!.addView(tvTitle)
        linear!!.addView(tvSubtitle)

        tvSubtitle!!.setSingleLine()
        tvSubtitle!!.ellipsize = TextUtils.TruncateAt.END
        tvSubtitle!!.setTextAppearance(
            context,
            R.style.TextAppearance_AppCompat_Widget_ActionBar_Subtitle
        )

        tvSubtitle!!.visibility = View.GONE

        val lp = Toolbar.LayoutParams(
            Toolbar.LayoutParams.WRAP_CONTENT,
            Toolbar.LayoutParams.WRAP_CONTENT
        )
        lp.gravity = Gravity.CENTER
        linear!!.layoutParams = lp

        addView(linear)
    }

}
